io_port.c

```diff
@@ -105,30 +77,127 @@ io_port_deregister(void)
                io_port.addr + io_port.size - 1);
 
     io_port.addr = IO_PORT_UNMAPPED;
+
+    free(io_port.readl);
+    free(io_port.writel);
 }
 
 void
 io_port_write(uint64_t addr, uint64_t size, uint32_t val)
 {
+    uint64_t port_index;
+    uint64_t alignment;
+
     assert(io_port.enable && io_port.addr <= addr
            && addr < (io_port.addr + io_port.size));
 
     addr -= io_port.addr;
+    alignment = addr % sizeof(uint32_t);
+
+    if (size == 0) {
+       DBG("Zero sized write to %" PRIu64 "\n", addr);
+    }
 
-    if (size == 4)
-        io_port.writel(addr, val);
-    else
-        DBG("Expected size 4. Got %" PRIu64 ".\n", size);
+    if (size + alignment <= sizeof(uint32_t)) {
+        port_index = addr / sizeof(uint32_t);
+        if (io_port.writel[port_index] != NULL)
+           io_port.writel[port_index](addr - port_index * sizeof(uint32_t), size, val);
+        else
+           DBG("Write to offset %" PRIu64 " not supported.\n", addr);
+    } else {
+        DBG("Write misaligned with ports. Address %" PRIu64 " Size %" PRIu64 ".\n", addr, size);
+    }
+}
+
+uint64_t
+io_port_read(uint64_t addr, uint64_t size)
+{
+    uint32_t port_index;
+    uint64_t alignment;
+
+    assert(io_port.enable && io_port.addr <= addr
+           && addr < (io_port.addr + io_port.size));
+
+    addr -= io_port.addr;
+    alignment = addr % sizeof(uint32_t);
+
+    if (size == 0) {
+       DBG("Zero sized read from %" PRIu64 "\n", addr);
+    }
+
+    if (size + alignment <= sizeof(uint32_t)) {
+        port_index = addr / sizeof(uint32_t);
+
+        if (io_port.readl[port_index] != NULL) {
+            return (uint64_t) io_port.readl[port_index](addr - port_index * sizeof(uint32_t), size);
+        } else {
+            DBG("Read from offset %" PRIu64 "not supported.\n", addr);
+            return 0;
+        }
+    } else {
+        DBG("Read misaligned with ports. Address %" PRIu64 " Size %" PRIu64 ".\n", addr, size);
+    }
+
+    return 0;
+}
+
+
+static int
+get_port(uint32_t address, uint32_t *port) {
+
+    if (address < io_port.addr  || address > io_port.addr + io_port.size - sizeof(uint32_t)) {
+        ERR("Failed to register io port handler, address out of range: %016"PRIx32"\n", address);
+        return -1;
+    }
+    if ((address - io_port.addr) & 0x3) {
+        ERR("Failed to register io port handler, address not aligned: %016"PRIx32"\n", address);
+        return -1;
+    }
+    *port = (address - io_port.addr) >> 2;
+    return 0;
+}
+
+int
+register_io_port_readl_handler(uint64_t address, readl_callback_t callback) {
+    uint32_t port;
+    int r;
+    r = get_port(address, &port);
```

Comment:
As `get_port()` simply either returns success or failure, it's probably better for it to return a bool.

Then you can simplify this to `if (!get_port(address, &port)) ...`

The same goes for the return types of `register_io_port_readl_handler` and `register_io_port_writel_handler`.